alleyinteractive/coral-talk
# CSS import overrides

## What does this PR do?

_As a developer implementing Talk, I'd like to be able to override CSS module import paths so I can have more control over Talk's look._

Using Talk's admin option to load a CSS file is great for minor tweaks, but bigger modifications to the styling of the UI require a lot of style overrides and `!important`. 

This PR allows a developer to create a `css-overrides.json` file in the Talk root directory. This file contains an array of resources, each with relative paths from the Talk root to the "old" and "new" resources.

If `css-overrides.json` exists, is valid JSON, and is not empty, Talk's Webpack config will use the [`NormalModuleReplacementPlugin`](https://webpack.js.org/plugins/normal-module-replacement-plugin/) to append the new resource to the original before importing as a CSS module. _Both the original and replacement resources must be a CSS files._

### Example

`css-overrides.json`:
```
[
  {
    "oldPath": "client/coral-embed-stream/src/components/Example.css",
    "newPath": "plugins/my-cool-plugin/styles/ExampleOverride.css"
  }
]
```

`Example.css`:
```
.anExample {
  color: blue;
  margin: 15px;
}
```

`ExampleOverride.css`:
```
.anExample {
  color: blue;
}
```


Now `Comment.js` will [import](https://github.com/coralproject/talk/blob/master/client/coral-embed-stream/src/components/Comment.js#L11) its `styles` object as `ExampleOverride.css` appended to `Example.css`:

```
.anExample {
  color: blue;
  margin: 15px;
}

.anExample {
  color: orange;
}
```

When the CSS modules are rendered, `<div className={styles.anExample}>` will inherit:
```
.anExample {
  margin: 15px;
  color: orange;
}
```

## How do I test this PR?

1. Create a file `css-overrides.json` in the Talk root using the above JSON snippet, replacing `Example` with `Comment`.
1. In the new `CommentOverride.css`, add `.content { color: orange }`
1. `yarn build`
1. The text of comments in the embed stream should be orange.
1. `git status` in the Talk root dir should not show any tracked changes.

Edge-case handling:

* Edit `css-overrides.json` so that it is no longer valid JSON
* Edit `oldPath` so that it ends with `Comment.js`
* Edit `newPath` so that it ends with `CommentOverride.js`

Then rebuild. In all cases, the Webpack build should complete without error but the styles will be inherited only from the original CSS file, `coral-embed-stream/src/components/Comment.css`.
